pkg/slayers/scion.go

```diff
@@ -598,48 +604,103 @@ func scionNextLayerTypeL4(t L4ProtocolType) gopacket.LayerType {
 }
 
 // SetDstAddr sets the destination address and updates the DstAddrType field accordingly.
-func (s *SCION) SetDstAddr(dst addr.Host) error {
+// @ requires  acc(&s.RawDstAddr)
+// @ requires  acc(&s.DstAddrType)
+// @ ensures   IsHostTypeSVC(dst) ==> res == nil
+// @ ensures   acc(&s.RawDstAddr) && acc(&s.DstAddrType)
+// @ ensures   res != nil ==> res.ErrorMem()
+// @ ensures   res == nil ==> IsHostTypeIP(dst) || IsHostTypeSVC(dst)
+// @ ensures   res == nil ==> sl.Bytes(s.RawDstAddr, 0, len(s.RawDstAddr))
+// @ decreases
+func (s *SCION) SetDstAddr(dst addr.Host) (res error) {
 	var err error
 	s.DstAddrType, s.RawDstAddr, err = PackAddr(dst)
 	return err
 }
 
 // SetSrcAddr sets the source address and updates the DstAddrType field accordingly.
-func (s *SCION) SetSrcAddr(src addr.Host) error {
+// @ requires  acc(&s.RawSrcAddr)
+// @ requires  acc(&s.SrcAddrType)
+// @ ensures   IsHostTypeSVC(src) ==> res == nil
+// @ ensures   acc(&s.RawSrcAddr) && acc(&s.SrcAddrType)
+// @ ensures   res != nil ==> res.ErrorMem()
+// @ ensures   res == nil ==> IsHostTypeIP(src) || IsHostTypeSVC(src)
+// @ ensures   res == nil ==> sl.Bytes(s.RawSrcAddr, 0, len(s.RawSrcAddr))
+// @ decreases
+func (s *SCION) SetSrcAddr(src addr.Host) (res error) {
 	var err error
 	s.SrcAddrType, s.RawSrcAddr, err = PackAddr(src)
 	return err
 }
 
-func ParseAddr(addrType AddrType, raw []byte) (addr.Host, error) {
+// @ requires addrType == T4Svc ==> len(raw) >= HostLenSVC
+// @ requires acc(sl.Bytes(raw, 0, len(raw)), R15)
+// @ ensures  err == nil ==> IsHostTypeIP(res) || IsHostTypeSVC(res)
+// @ ensures  acc(sl.Bytes(raw, 0, len(raw)), R15)
+// @ ensures  err != nil ==> err.ErrorMem()
+// @ decreases
+func ParseAddr(addrType AddrType, raw []byte) (res addr.Host, err error) {
 	switch addrType {
 	case T4Ip:
-		var raw4 [4]byte
-		copy(raw4[:], raw)
+		var raw4 /*@ @ @*/ [4]byte
+		// @ unfold acc(sl.Bytes(raw, 0, len(raw)), R15)
+		copy(raw4[:], raw /*@, R15 @*/)
+		// @ fold acc(sl.Bytes(raw, 0, len(raw)), R15)
 		return addr.HostIP(netip.AddrFrom4(raw4)), nil
 	case T4Svc:
+		// @ unfold acc(sl.Bytes(raw, 0, len(raw)), R15)
 		svc := addr.SVC(binary.BigEndian.Uint16(raw[:2]))
+		// @ fold acc(sl.Bytes(raw, 0, len(raw)), R15)
 		return addr.HostSVC(svc), nil
 	case T16Ip:
-		var raw16 [16]byte
-		copy(raw16[:], raw)
+		var raw16 /*@ @ @*/ [16]byte
+		// @ unfold acc(sl.Bytes(raw, 0, len(raw)), R15)
+		copy(raw16[:], raw /*@, R15 @*/)
+		// @ fold acc(sl.Bytes(raw, 0, len(raw)), R15)
 		return addr.HostIP(netip.AddrFrom16(raw16)), nil
 	}
 	return addr.Host{}, serrors.New("unsupported address type/length combination",
 		"type", addrType, "len", addrType.Length())
 }
 
-func PackAddr(host addr.Host) (AddrType, []byte, error) {
+// @ ensures   IsHostTypeSVC(host) ==> err == nil
+// @ ensures   err == nil ==> IsHostTypeIP(host) || IsHostTypeSVC(host)
+// @ ensures   err != nil ==> err.ErrorMem()
+// @ ensures   err == nil ==> sl.Bytes(b, 0, len(b))
```

Comment:
Here, and everywhere else where this function is called, we may even make this postcondition simpler by removing the condition (we can fold `sl.Bytes(b, 0, len(b))` when `b` is `nil`). This might be preferable, given that (1) avoiding an impure implication (i.e., an implication with a resource) might remove the number of branches that silicon explores, (2) conceptually, it is still fine for the client to read all (0) bytes of the the returned slice.

Whenever possible, we should try to avoid impure implications

```suggestion
// @ ensures   sl.Bytes(b, 0, len(b))
```

Reply:
I see. I did not know that. Thank you!

I adapted the specs accordingly.

Done.